### PR TITLE
SCRUM-903: CHEBITerm CRUD

### PR DESCRIPTION
**Pull request by abecerra**
No bulk upload, UI yet

### CODE REVIEW

**Comment by oblodgett** [comment]
I assume the build part of this is coming?

**Comment by abecerra** [comment]
> I assume the build part of this is coming?

Yes!

**Comment by oblodgett** [comment]
Sorry I meant "BULK".

File: src/main/cliapp/src/components/ontologies/CHEBIOntologyComponent.js [hunk 1]
```diff
@@ -67,7 +67,7 @@ export const CHEBIOntologyComponent = () => {
   return (
       <div>
         <div className="card">
-            <h3>ZFA Table</h3>
+            <h3>CHEBI Table</h3>
```
Comment (markquintontulloch) [inline src/main/cliapp/src/components/ontologies/CHEBIOntologyComponent.js:70]: This should be ChEBI with a lower case 'h'

File: src/main/java/org/alliancegenome/curation_api/interfaces/crud/ontology/CHEBITermRESTInterface.java [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+import org.alliancegenome.curation_api.model.entities.ontology.CHEBITerm;
+import org.eclipse.microprofile.openapi.annotations.tags.Tag;
+
+import javax.ws.rs.Consumes;
+import javax.ws.rs.Path;
+import javax.ws.rs.Produces;
+import javax.ws.rs.core.MediaType;
+
+@Path("/chebiterm")
+@Tag(name = "Ontology - CHEBI")
```
Comment (markquintontulloch) [inline src/main/java/org/alliancegenome/curation_api/interfaces/crud/ontology/CHEBITermRESTInterface.java:13]: Should be ChEBI

File: src/main/java/org/alliancegenome/curation_api/interfaces/bulk/ontology/CHEBITermBulkRESTInterface.java [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+
+import org.alliancegenome.curation_api.base.BaseOntologyTermBulkRESTInterface;
+import org.eclipse.microprofile.openapi.annotations.tags.Tag;
+
+import javax.ws.rs.Path;
+import javax.ws.rs.Produces;
+import javax.ws.rs.core.MediaType;
+
+@Path("/chebiterm/bulk")
+@Tag(name = "Ontology - CHEBI")
```
Comment (markquintontulloch) [inline src/main/java/org/alliancegenome/curation_api/interfaces/bulk/ontology/CHEBITermBulkRESTInterface.java:11]: Should be ChEBI

**Review by markquintontulloch — CHANGES_REQUESTED** [review]
Looks good, just the 'ChEBI' instead of 'CHEBI' for display.  Before we pull this in we need to either change the 'name' column type to TEXT in the OntologyTerm entity class or only load the first 255 characters, otherwise the bulk load will fail.  I'm inclined to do the former and pull this in with the changes for SCRUM-510 in the next sprint, just in case the first 255 characters isn't enough to distinguish between entries in some cases.

File: src/main/cliapp/src/App.js [hunk 4]
```diff
@@ -131,6 +131,7 @@ const App = () => {
         {
             label: 'Ontologies', icon: 'pi pi-fw pi-sitemap',
             items: [
+                { label: 'CHEBI Ontology', icon: 'pi pi-fw pi-home', to: '/ontology/chebi' },
```
Comment (markquintontulloch) [inline src/main/cliapp/src/App.js:134]: For consistency, should be:
`{ label: 'Chemical Entities of Biological Interest (ChEBI)', icon: 'pi pi-fw pi-home', to: '/ontology/chebi' },`

File: src/main/cliapp/src/components/ontologies/CHEBIOntologyComponent.js (hunk 1 above)
Comment (abecerra) [reply src/main/cliapp/src/components/ontologies/CHEBIOntologyComponent.js]: Done

File: src/main/java/org/alliancegenome/curation_api/interfaces/crud/ontology/CHEBITermRESTInterface.java (hunk 2 above)
Comment (abecerra) [reply src/main/java/org/alliancegenome/curation_api/interfaces/crud/ontology/CHEBITermRESTInterface.java]: Done

File: src/main/java/org/alliancegenome/curation_api/interfaces/bulk/ontology/CHEBITermBulkRESTInterface.java (hunk 3 above)
Comment (abecerra) [reply src/main/java/org/alliancegenome/curation_api/interfaces/bulk/ontology/CHEBITermBulkRESTInterface.java]: Done!

**Comment by abecerra** [comment]
> Looks good, just the 'ChEBI' instead of 'CHEBI' for display. Before we pull this in we need to either change the 'name' column type to TEXT in the OntologyTerm entity class or only load the first 255 characters, otherwise the bulk load will fail. I'm inclined to do the former and pull this in with the changes for SCRUM-510 in the next sprint, just in case the first 255 characters isn't enough to distinguish between entries in some cases.

I tested with @Lob annotation on OntologyTerm.name and hibernate sets the field as TEXT in postgres. 

However, the size of chebi.owl (634MB) crashes the backend for lack  of Java heap space

`2021-12-06 14:48:01,460 ERROR [io.qua.ver.htt.run.QuarkusErrorHandler] (executor-thread-0) HTTP Request to /api/chebiterm/bulk/owl failed, error id: 77ffbfed-0a23-4e5d-9a28-0c3a3da4422a-1: java.lang.OutOfMemoryError: Java heap space
        at java.base/java.util.Arrays.copyOf(Arrays.java:3745)
        at java.base/java.io.ByteArrayOutputStream.grow(ByteArrayOutputStream.java:120)
        at java.base/java.io.ByteArrayOutputStream.ensureCapacity(ByteArrayOutputStream.java:95)
        at java.base/java.io.ByteArrayOutputStream.write(ByteArrayOutputStream.java:156)
        at org.jboss.resteasy.plugins.providers.ProviderHelper.readString(ProviderHelper.java:79)
        at org.jboss.resteasy.plugins.providers.StringTextStar.readFrom(StringTextStar.java:45)
        at org.jboss.resteasy.plugins.providers.StringTextStar.readFrom(StringTextStar.java:27)
        at org.jboss.resteasy.core.interception.jaxrs.AbstractReaderInterceptorContext.readFrom(AbstractReaderInterceptorContext.java:101)
        at org.jboss.resteasy.core.interception.jaxrs.ServerReaderInterceptorContext.readFrom(ServerReaderInterceptorContext.java:63)
        at org.jboss.resteasy.core.interception.jaxrs.AbstractReaderInterceptorContext.proceed(AbstractReaderInterceptorContext.java:80)
        at org.jboss.resteasy.core.MessageBodyParameterInjector.inject(MessageBodyParameterInjector.java:213)
        at org.jboss.resteasy.core.MethodInjectorImpl.injectArguments(MethodInjectorImpl.java:95)
        at org.jboss.resteasy.core.MethodInjectorImpl.invoke(MethodInjectorImpl.java:128)
        at org.jboss.resteasy.core.ResourceMethodInvoker.internalInvokeOnTarget(ResourceMethodInvoker.java:660)
        at org.jboss.resteasy.core.ResourceMethodInvoker.invokeOnTargetAfterFilter(ResourceMethodInvoker.java:524)
        at org.jboss.resteasy.core.ResourceMethodInvoker.lambda$invokeOnTarget$2(ResourceMethodInvoker.java:474)
        at org.jboss.resteasy.core.ResourceMethodInvoker$$Lambda$1750/0x0000000840da7440.get(Unknown Source)
        at org.jboss.resteasy.core.interception.jaxrs.PreMatchContainerRequestContext.filter(PreMatchContainerRequestContext.java:364)
        at org.jboss.resteasy.core.ResourceMethodInvoker.invokeOnTarget(ResourceMethodInvoker.java:476)
        at org.jboss.resteasy.core.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:434)
        at org.jboss.resteasy.core.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:408)
        at org.jboss.resteasy.core.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:69)
        at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:492)
        at org.jboss.resteasy.core.SynchronousDispatcher.lambda$invoke$4(SynchronousDispatcher.java:261)
        at org.jboss.resteasy.core.SynchronousDispatcher$$Lambda$1748/0x0000000840da7c40.run(Unknown Source)
        at org.jboss.resteasy.core.SynchronousDispatcher.lambda$preprocess$0(SynchronousDispatcher.java:161)
        at org.jboss.resteasy.core.SynchronousDispatcher$$Lambda$1749/0x0000000840da7040.get(Unknown Source)
        at org.jboss.resteasy.core.interception.jaxrs.PreMatchContainerRequestContext.filter(PreMatchContainerRequestContext.java:364)
        at org.jboss.resteasy.core.SynchronousDispatcher.preprocess(SynchronousDispatcher.java:164)
        at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:247)
        at io.quarkus.resteasy.runtime.standalone.RequestDispatcher.service(RequestDispatcher.java:73)
        at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler.dispatch(VertxRequestHandler.java:138)`

**Comment by abecerra** [comment]
It seems this space problem only occurs in my machine for having only 8GB.

**Comment by oblodgett** [comment]
@abecerra what is holding this up? just that you can't load it on your machine? We could try to push it out and see what happens with Alpha?

**Comment by abecerra** [comment]
> @abecerra what is holding this up? just that you can't load it on your machine? We could try to push it out and see what happens with Alpha?

Yes, only on my machine I suppose, as @markquintontulloch  loaded CheBI just fine.
When we merge: https://github.com/alliance-genome/agr_curation/pull/170 we can merge this pr

**Review by oblodgett — APPROVED** [review]

**Review by oblodgett — APPROVED** [review]

**Review by markquintontulloch — APPROVED** [review]

**Pull request merged**